src/rez/package_test.py

```diff
@@ -158,45 +220,106 @@ def print_header(txt, *nargs):
+                # This works because RequirementList merges requirements.
+                #
+                if RequirementList(variant.variant_requires) != reqlist:
+                    continue
 
             # show progress
             if self.verbose:
-                print_header(
+                self._print_header(
                     "\nTest: %s\nPackage: %s\n%s\n",
-                    test_name, variant.uri, '-' * 80)
+                    test_name, variant.uri, '-' * 80
+                )
+
+            # add requirements to force the current variant to be resolved.
+            # TODO this is not perfect, and will need to be updated when
+            # explicit variant selection is added to rez (this is a new
+            # feature). Until then, there's no guarantee that we'll resolve to
+            # the variant we want, so we take that into account here.
+            #
+            requires.extend(map(str, variant.variant_requires))
 
-            # create test env
-            key = tuple(requires)
-            context = self.contexts.get(key)
+            # create test runtime env
+            context = self._get_context(requires)
 
             if context is None:
-                if self.verbose:
-                    print_header("Resolving test environment: %s\n",
-                                 ' '.join(map(quote, requires)))
+                self._set_test_result(
+                    test_name,
+                    variant,
+                    "skipped",
+                    "The current environment does not meet test requirements"
+                )
+
+                print("(Skipped)")
+                continue
 
-                context = ResolvedContext(package_requests=requires,
-                                          package_paths=self.package_paths,
-                                          buf=self.stdout,
-                                          timestamp=self.timestamp,
-                                          **self.context_kwargs)
+            if not context.success:
+                self._set_test_result(
+                    test_name,
+                    variant,
+                    "skipped",
+                    "The test environment failed to resolve"
+                )
+                continue
 
-                if not context.success:
-                    context.print_info(buf=self.stderr)
-                    raise PackageTestError(
-                        "Cannot run test '%s' of package %s: the environment "
-                        "failed to resolve" % (test_name, variant.uri))
+            # check that this has actually resolved the variant we want
+            resolved_variant = context.get_resolved_package(package.name)
+            assert resolved_variant
+
+            if resolved_variant.handle != variant.handle:
+                print_warning(
+                    "Could not resolve environment for this variant. This is a "
```

Comment:
Yeah agreed, will add the source variant to the output.

Not sure what you mean by the latter. This can happen in any case where you're attempting to resolve an env to contain a specific variant of a package.